pkg/client/openstack/kluster/client.go

```diff
@@ -190,94 +202,249 @@ func (c *klusterClient) SetSecurityGroup(sgName, nodeID string) (err error) {
 	return secgroups.AddServer(c.ComputeClient, nodeID, sgName).ExtractErr()
 }
 
-func (c *klusterClient) EnsureKubernikusRuleInSecurityGroup(kluster *v1.Kluster) (created bool, err error) {
-	sgName := kluster.Spec.Openstack.SecurityGroupName
-	page, err := securitygroups.List(c.NetworkClient, securitygroups.ListOpts{Name: sgName}).AllPages()
+type portListOpts struct {
+	ProjectID      string `q:"project_id"`
+	SecurityGroups string `q:"security_groups"`
+}
+
+func (opts portListOpts) ToPortListQuery() (string, error) {
+	q, err := gophercloud.BuildQueryString(opts)
+	params := q.Query()
+	q = &url.URL{RawQuery: params.Encode()}
+	return q.String(), err
+}
+
+func getLoadbalancerSecurityGroup(kluster *v1.Kluster, client *gophercloud.ServiceClient) (*securitygroups.SecGroup, error) {
+	lbSgName := fmt.Sprintf(LoadBalancerSecurityGroupName, kluster.Spec.Name)
+	lbSgPage, err := securitygroups.List(client, securitygroups.ListOpts{Name: lbSgName}).AllPages()
 	if err != nil {
-		return false, fmt.Errorf("SecurityGroup %v not found: %s", sgName, err)
+		return nil, fmt.Errorf("Loadbalancer SecurityGroup %s listing error: %w", lbSgName, err)
 	}
 
-	if kluster.ClusterCIDR() == "" {
-		return false, errors.New("Cluster CIDR for kluster not set")
+	lbGroups, err := securitygroups.ExtractGroups(lbSgPage)
+	if err == nil && len(lbGroups) == 1 {
+		return &lbGroups[0], nil
 	}
 
-	groups, err := securitygroups.ExtractGroups(page)
+	return nil, err
+}
+
+func createLoadbalancerSecurityGroup(kluster *v1.Kluster, client *gophercloud.ServiceClient) (*securitygroups.SecGroup, error) {
+	sg, err := securitygroups.Create(client, securitygroups.CreateOpts{
+		Name:        fmt.Sprintf(LoadBalancerSecurityGroupName, kluster.Spec.Name),
+		Description: fmt.Sprintf(`Kubernikus: SecurityGroup containing LoadBalancer Self-IPs for cluster "%s"`, kluster.Spec.Name),
+	}).Extract()
 	if err != nil {
-		return false, err
+		return nil, err
 	}
 
-	if len(groups) != 1 {
-		return false, fmt.Errorf("More than one SecurityGroup with name %v found", sgName)
+	return sg, nil
+}
+
+func (c *klusterClient) EnsureLoadbalancerSecurityGroup(kluster *v1.Kluster, recorder record.EventRecorder) (created bool, err error) {
+	var lbGroup *securitygroups.SecGroup
+	lbGroup, err = getLoadbalancerSecurityGroup(kluster, c.NetworkClient)
+	if err != nil {
+		return false, err
 	}
 
-	udp := false
-	tcp := false
-	icmp := false
-	for _, rule := range groups[0].Rules {
-		if rule.Direction != string(rules.DirIngress) {
-			continue
+	if lbGroup == nil {
+		lbGroup, err = createLoadbalancerSecurityGroup(kluster, c.NetworkClient)
+		if err != nil || lbGroup == nil {
```

Comment:
not sure the `|| lbGroup == nil` is required. Looking at createLoadbalancerSecurityGroup it returns an error when the create call fails. I don't see how `err` and `lbGroup` can be nil at the same time.